From 8df3a7a2c49db4509aedd5f3cf1e3340cd9d7b82 Mon Sep 17 00:00:00 2001
From: Alan Coopersmith <alan.coopersmith@sun.com>
Date: Fri, 6 Nov 2009 16:36:45 -0800
Subject: [PATCH] Use /dev/urandom for xauth cookie generation

The previous comment about only Linux having it is a bit outdated,
it's common on most modern Unix-like systems now, and we can easily
fall back to the old method on systems without it.
---
 unix/vncserver |   25 +++++++++++++++++--------
 1 files changed, 17 insertions(+), 8 deletions(-)

diff --git a/unix/vncserver b/unix/vncserver
index 90ef0d2..28764cb 100755
--- a/unix/vncserver
+++ b/unix/vncserver
@@ -189,16 +189,25 @@ $vncPort = 5900 + $displayNumber;
 $desktopLog = "$vncUserDir/$host:$displayNumber.log";
 unlink($desktopLog);
 
-# Make an X server cookie - use as the seed the sum of the current time, our
-# PID and part of the encrypted form of the password.  Ideally we'd use
-# /dev/urandom, but that's only available on Linux.
-
-srand(time+$$+unpack("L",`cat $vncUserDir/passwd`));
-$cookie = "";
-for (1..16) {
+# Make an X server cookie - use /dev/urandom on systems that have it,
+# otherwise use perl's random number generator, seeded with the sum
+# of the current time, our PID and part of the encrypted form of the password.
+
+my $cookie = "";
+if (open(URANDOM, '<', '/dev/urandom')) {
+  my $randata;
+  if (sysread(URANDOM, $randata, 16) == 16) {
+    $cookie = unpack 'h*', $randata;
+  }
+  close(URANDOM);
+}
+if ($cookie eq "") {
+  srand(time+$$+unpack("L",`cat $vncUserDir/passwd`));
+  for (1..16) {
     $cookie .= sprintf("%02x", int(rand(256)) % 256);
+  }
 }
-    
+
 system("xauth -f $xauthorityFile add $host:$displayNumber . $cookie");
 system("xauth -f $xauthorityFile add $host/unix:$displayNumber . $cookie"); 
 
-- 
1.5.6.5

